code/platforms/mac/app.py

```diff
@@ -1,34 +1,66 @@
-from talon import Context, actions
+from talon import Context, actions, ui
+
 ctx = Context()
 ctx.matches = r"""
 os: mac
 """
 
+
 @ctx.action_class('app')
 class AppActions:
     def preferences():
         actions.key('cmd-,')
+
     def tab_close():
         actions.key('cmd-w')
-        #action(app.tab_detach):
+        # action(app.tab_detach):
         #  Move the current tab to a new window
+
     def tab_next():
         actions.key('cmd-alt-right')
+
     def tab_open():
         actions.key('cmd-t')
+
     def tab_previous():
         actions.key('cmd-alt-left')
+
     def tab_reopen():
         actions.key('cmd-shift-t')
+
     def window_close():
         actions.key('cmd-w')
+
     def window_hide():
         actions.key('cmd-m')
+
     def window_hide_others():
         actions.key('cmd-alt-h')
-    def window_next():
-        actions.key('cmd-`')
+
     def window_open():
         actions.key('cmd-n')
+
+    # Custom behavior to handle Mac desktop 'Spaces'
+    def window_next():
+        switch_window_by_offset_from_current(1)
+
+    # Custom behavior to handle Mac desktop 'Spaces'
     def window_previous():
-        actions.key('cmd-shift-`')
+        switch_window_by_offset_from_current(-1)
+
+
+def switch_window_by_offset_from_current(offset):
+    active_window = ui.active_window()
+    windows = ui.active_app().windows()
+
+    # Handle case where empty windows with no title show up in the results
+    windows = list(filter(lambda local_window: len(local_window.title) > 0, windows))
+
+    # Sort by title as the windows get reordered on every switch and we want to switch windows in a fixed order
+    windows.sort(key=lambda local_window: local_window.title)
```

Comment:
In theory there's an edge case where two windows have the same title which makes it hard to determine which one is the active one (since the other params like pid and app are the same). 

I'm hoping this is solved in terms of exact matching by using the `is` keyword, but I don't have a good way to solve it for the sorting

Maybe @lunixbochs is able to add an extra identifier to the `ui.window` objects (or is there already an extra ID I'm missing)